sdk/src/main/scala/com/horizen/account/api/rpc/service/EthService.scala

```diff
@@ -59,57 +54,75 @@ class EthService(val stateView: AccountStateView, val nodeView: CurrentView[Acco
     if (blockId == null) return null
     val block = nodeView.history.getBlockById(blockId)
     if (block.isEmpty) return null
-    val transactions = getTransactionsFromBlock(block.get())
-    val txList: ListBuffer[Any] = ListBuffer()
-    transactions.foreach(transaction => {
-      if (hydratedTx) {
-        val receipt = stateView.getTransactionReceipt(Numeric.hexStringToByteArray(Numeric.cleanHexPrefix(transaction.id)))
-        if (!receipt.isEmpty) txList.append(new EthereumTransactionView(receipt.get, transaction))
-      } else txList.append(Numeric.prependHexPrefix(transaction.id))
-    })
-    new EthereumBlock(Numeric.prependHexPrefix(Integer.toHexString(nodeView.history.getBlockHeight(blockId).get())), Numeric.prependHexPrefix(blockId), txList.toList.asJava, block.get())
-  }
+    // TODO: why does this not compile?
+    //val transactions = block.get().transactions.collect { case tx: EthereumTransaction => tx }
+    val transactions = block.get().transactions.filter {
+      _.isInstanceOf[EthereumTransaction]
+    } map {
+      _.asInstanceOf[EthereumTransaction]
+    }
+    new EthereumBlock(
+      Numeric.prependHexPrefix(Integer.toHexString(nodeView.history.getBlockHeight(blockId).get())),
+      Numeric.prependHexPrefix(blockId),
+      if (!hydratedTx) {
+        transactions.map(tx => Numeric.prependHexPrefix(tx.id)).toList.asJava
+      } else {
+        transactions.flatMap(tx => stateView.getTransactionReceipt(Numeric.hexStringToByteArray(tx.id)) match {
+          case Some(receipt) => Some(new EthereumTransactionView(receipt, tx))
+          case None => None
+        }).toList.asJava
+      },
+      block.get())
+  }
+
+  private def doCall(params: TransactionArgs, tag: Quantity) = {
+    using(getStateViewFromBlockById(getBlockIdByTag(tag.getValue))) { stateDbRoot =>
+      if (stateDbRoot == null) {
+        throw new IllegalArgumentException(s"Unable to get state for given tag: ${tag.getValue}")
+      }
+      val message = params.toMessage
+
+      // TODO: refactor to reuse parts of AccountStateView.applyTransaction?
+      val messageProcessors = Seq(
```

Comment:
I would prefer to directly use the sequence of message processors of the current stateView, but that is not accessible currently. Suggestions?

Reply:
Let's implement `staticCall` method in the AccountStateView class, which has all needed elements inside.
EthService should just decide at which point to open the view and pass the message

Reply:
AccountStateView now has `applyMessage` to be closer to geth naming.